lib/GADS.pm

```diff
@@ -3024,6 +3025,89 @@ prefix '/:layout_name' => sub {
+        my $layout = var('layout') or pass;
+
+        forwardHome({ danger => "You do not have permission to manage deleted records"}, '')
+            unless $layout->user_can("purge");
+
+        my $user        = logged_in_user;
+        my $view        = current_view($user, $layout);
+
+        my $records = GADS::Records->new(
+            user                => $user,
+            layout              => $layout,
+            schema              => schema,
+            include_children    => 1,
+            view                => $view,
+        );
+
+        my $params = +{};
+        my @record_data;
+
+        if(param 'submit') {
+            my @columns = body_parameters->get_all('column_id');
+            $records->columns(\@columns);
+        } elsif(param 'purge') {
+            my @columns = body_parameters->get_all('column_id');
+            $records->columns(\@columns);
+        }
+
+        unless(param 'purge') {
+            while (my $record = $records->single) {
+                foreach my $column (@{$record->columns_render}) {
+                    next if $column->internal;
+
+                    my $clone = $record->clone;
+                    my $count;
+                    my $current_id = $record->current_id;
+                    $clone->find_chronology_id($current_id);
+                    foreach my $cron (@{$clone->chronology}) {
+                        $count = scalar(grep {$_->{name} eq $column->name}  @{$cron->{changed}}) + 1;
+                    }
+                    
+                    my @pmc= $record->presentation_map_columns(columns=> [$column]);
+                    push @record_data, {
+                        col_id => $column->id,
+                        name => $column->name,
+                        id => $record->record->{id},
+                        value => $pmc[0]->{data}->{value}, # THIS SHOULD BE THE VALUE OF THE GIVEN RECORD/COLUMN
+                        count => $count || 0,
+                    };
+                }
+            }
+        }
+
+        if(param 'submit') {
+            $params->{records} = \@record_data;
+            my @mapped_ids = map {$_->{col_id}} @record_data;
+            my @mapped_records = uniq map {$_->{id}} @record_data;
+            $params->{column_list} = \@mapped_ids;
+            $params->{record_list} = \@mapped_records;
+            return template 'historic_purge/confirm' => $params;
+        } elsif(param 'purge') {
+            my @columns = body_parameters->get_all('columns') or error __"Unable to get columns";
+            my @rids = body_parameters->get_all('column_id') or error __"Unable to get records";
+            my $total = 0;
+            foreach my $r (@rids) {
+                my $record = schema->resultset('Record')->find($r);
+                my $current_id = $record->current_id;
+                my $current = schema->resultset('Current')->find($current_id);
+                $current->historic_purge(@columns);
```

Comment:
This doesn't look quite right. It should be possible to obtain the current_id directly?